L69-cpp-core-event-manager.md

```diff
@@ -0,0 +1,282 @@
+
+The EventManager interface aims at simplifying the usage and maintenance of gRPC event framework. It allows users of gRPC to write their own event management implementation to be used by gRPC for specific needs. The project also paves the way for supporting features such as the C++ callback API in gRPC (See proposal [gRFC L67: C++ callback-based asynchronous API](https://github.com/grpc/proposal/pull/180)).
+
+## Proposal
+The following event manager interfaces will be available in the gRPC core.
+
+```cpp
+// An external handle to an EventManager task, which can be used to cancel the 
+// task. The EventManager task is added via RunFunction.
+struct TaskKey;
+
+// An external handle to an EventManager alarm, which can be used to cancel the
+// alarm. The EventManager alarm is added via AddAlarmIn.
+struct AlarmKey;
+
+// Scheduling priority for closures and alarms added into the EventManager.
+enum Priority {
+  PRIORITY_MIN = 0,
+  PRIORITY_LOW = 0,
+  PRIORITY_MED = 1,
+  PRIORITY_HIGH = 2,
+  PRIORITY_IMMEDIATE = 3,
+  PRIORITY_SYSTEM = 4,
+  PRIORITY_MAX = 4,
+  NUM_PRIORITIES = 5
+};
+
+// A top-level base class of EventManagerInterface.
+class BaseEventManagerInterface {
+ public:
+  virtual ~BaseEventManagerInterface();
+
+  // Adds a new one-shot alarm, which will be scheduled at the deadline to run the 
+  // closure with true. The deadline here is an absolute time. The AlarmKey pointer 
+  // is an output parameter which can be used to cancel the alarm before it fires. 
+  // Note that the closure should not be blocking or long-running, as it would 
+  // block the worker thread in the EventManager from polling the file 
+  // descriptors. In case that the alarm is cancelled successfully in 
+  // ConsistentCancel, the closure will be run with false in the calling thread of 
+  // ConsistentCancel. Callers of ConsistentCancel need to be cautious to avoid 
+  // deadlock.
+  virtual void ScheduleAlarmIn(gpr_timespec deadline, std::function<void(bool)> c, Priority priority, AlarmKey* key = nullptr) = 0;
+
+  // Adds a closure to the EventManager, which will be run at the earliest possible 
+  // time. The TaskKey pointer is an output parameter which can be used to cancel 
+  // the closure at a later time (currently not supported in 
+  // BaseEventManagerInterface). Note that the closure should not be blocking or 
+  // long-running, as it would block the worker thread in the EventManager from 
+  // polling the file descriptors.
+  virtual void RunFunction(std::function<void()> handler, Priority priority, TaskKey* key = nullptr) = 0;
+
+  struct Task {
+    std::function<void()> handler;
+    Priority priority;
+    TaskKey* key;
+    int internal_added_queue;
```

Comment:
What is the `internal_added_queue` member for?

Reply:
I've added some comment to clarify.

Reply:
Gotcha. Should this be a `void*` so that the implementation can use whatever "handle" type is most natural for it? E.g., for an IOCP based event manager, that could be a `HANDLE`, which won't fit in an `int`.

Reply:
Based on the same compatibility argument above, we will probably not make it a `void*`.

Reply:
That's really quite unfortunate. Is there really no way to make this pointer sized?

On Windows, the handles/descriptors for the OS primitives that would likely be used in an EM implementation, like an IO Completion Port `HANDLE`, a `RIO_CQ`, a WinSock `SOCKET`, and a file's `HANDLE` are 64-bits wide when targeting amd64. If this is only an `int` (32-bits wide when targeting amd64), then such implementations will have to have a mapping table so they can expose an `int` key.

libdispatch/Grand Central Dispatch similarly has pointer-sized handles/descriptors, like `dispatch_queue_t` which is typedef for a pointer to an internal struct.

(I'm assuming that the Google EM uses this field to store a file descriptor that is associated with the task queue. If not, knowing how it's used could allay my concerns.)